source/common/upstream/upstream_impl.h

```diff
@@ -324,6 +324,13 @@ class ClusterImplBase : public Cluster,
   // Upstream::Cluster
   ClusterInfoConstSharedPtr info() const override { return info_; }
   const Outlier::Detector* outlierDetector() const override { return outlier_detector_.get(); }
+  void setInitializedCb(std::function<void()> callback) override {
+    if (initialization_started_) {
+      callback();
```

Comment:
I think this should be if (initialization_started_ && pending_initialize_health_checks_ == 0) otherwise if you initialize() then set the callback you get the callback before health checking is complete.

Reply:
yup thanks, not spending enough time on this. will fix.